src/Learning/wls-linear-learn.jl

```diff
@@ -0,0 +1,102 @@
+# Weighted least squares functions ############################################
+
+"""
+function learn!(
+    lp::UnivariateLinearProblem,
+    ws::Vector,
+    int::Bool
+)
+
+Fit energies using weighted least squares.
+"""
+function learn!(
+    lp::UnivariateLinearProblem,
+    ws::Vector,
+    int::Bool
+)
+    @views B_train = reduce(hcat, lp.iv_data)'
+    @views e_train = lp.dv_data
+
+    # Calculate A and b.
+    if int
+        int_col = ones(size(B_train, 1))
+        @views A = hcat(int_col, B_train)
+    else
+        @views A = B_train
+    end
+    @views b = e_train
+
+    # Calculate coefficients β.
+    Q = Diagonal(ws[1] * ones(length(e_train)))
+    βs = (A'*Q*A) \ (A'*Q*b)
+    
+    # Update lp.
+    if int
+        copyto!(lp.β0, [βs[1]])
+        copyto!(lp.β, βs[2:end])
+    else
+        copyto!(lp.β, βs)
+    end
+    
+end
+
+"""
+function learn!(
+    lp::CovariateLinearProblem,
+    ws::Vector,
+    int::Bool
+)
+
+Fit energies and forces using weighted least squares.
+"""
+function learn!(
```

Comment:
At some point it would be important to resolve differences in how the CovariateLinearProblem is solved. In OLS, it is solved via optimization, while in WLS it is solved as a linear least squares problem.

Reply:
I completely agree, and I think we can revisit this topic in a follow-up to this PR.